File: medium/3121-count-the-number-of-special-characters-ii/solution.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        """
        3121. Count the Number of Special Characters II
        Time: O(n)
        Space: O(1)
        """
        n = len(word)
        last_lowercase = [-1] * 26
        first_uppercase = [n] * 26

        for index, char in enumerate(word):
            if char.islower():
                letter = ord(char) - ord("a")
                last_lowercase[letter] = index
            else:
                letter = ord(char) - ord("A")
                first_uppercase[letter] = min(first_uppercase[letter], index)

        answer = 0
        for letter in range(26):
            if (
                last_lowercase[letter] != -1
                and first_uppercase[letter] != n
                and last_lowercase[letter] < first_uppercase[letter]
            ):
                answer += 1

        return answer

File: medium/3121-count-the-number-of-special-characters-ii/solution.java
/**
 * 3121. Count the Number of Special Characters II
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int numberOfSpecialChars(String word) {
        int n = word.length();
        int[] lastLowercase = new int[26];
        int[] firstUppercase = new int[26];

        for (int letter = 0; letter < 26; letter++) {
            lastLowercase[letter] = -1;
            firstUppercase[letter] = n;
        }

        for (int index = 0; index < n; index++) {
            char current = word.charAt(index);
            if (Character.isLowerCase(current)) {
                int letter = current - 'a';
                lastLowercase[letter] = index;
            } else {
                int letter = current - 'A';
                firstUppercase[letter] = Math.min(firstUppercase[letter], index);
            }
        }

        int answer = 0;
        for (int letter = 0; letter < 26; letter++) {
            if (
                lastLowercase[letter] != -1
                    && firstUppercase[letter] != n
                    && lastLowercase[letter] < firstUppercase[letter]
            ) {
                answer++;
            }
        }

        return answer;
    }
}
